File: src/main/java/org/lanternpowered/server/network/message/HandlerMessage.java
/*
 * This file is part of LanternServer, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://github.com/LanternPowered>
 * Copyright (c) Contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.server.network.message;

import static com.google.common.base.Preconditions.checkNotNull;

import org.lanternpowered.server.network.message.handler.Handler;

public final class HandlerMessage<M extends Message> implements Message {

    private final M message;
    private final Handler<? super M> handler;

    public HandlerMessage(M message, Handler<? super M> handler) {
        this.message = checkNotNull(message, "message");
        this.handler = checkNotNull(handler, "handler");
    }

    public M getMessage() {
        return this.message;
    }

    public Handler<? super M> getHandler() {
        return this.handler;
    }
}

File: src/main/kotlin/org/lanternpowered/server/network/message/HandlerMessage.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.network.message

import org.lanternpowered.server.network.NettyThreadOnlyHelper
import org.lanternpowered.server.network.message.handler.Handler

data class HandlerMessage<M : Message> @JvmOverloads constructor(
        val message: M,
        val handler: Handler<in M>,
        val handleThread: HandleThread = getDefaultHandleThread(handler)
) : Message {

    enum class HandleThread {
        /**
         * The message is handled to the main (sync) server thread.
         */
        MAIN,
        /**
         * The message is handled on the netty thread.
         */
        NETTY,
        /**
         * The message is handled on a async thread.
         */
        ASYNC,
    }
}

fun getDefaultHandleThread(handler: Handler<*>): HandlerMessage.HandleThread {
    return if (NettyThreadOnlyHelper.isHandlerNettyThreadOnly(handler::class.java)) {
        HandlerMessage.HandleThread.NETTY
    } else {
        HandlerMessage.HandleThread.MAIN
    }
}
